clang/lib/Sema/Sema.cpp

```diff
@@ -1703,7 +1703,9 @@ class DeferredDiagnosticsEmitter
   void visitUsedDecl(SourceLocation Loc, Decl *D) {
     if (S.LangOpts.SYCLIsDevice && ShouldEmitRootNode) {
       if (auto *VD = dyn_cast<VarDecl>(D)) {
-        if (!S.checkAllowedSYCLInitializer(VD)) {
+        if (!S.checkAllowedSYCLInitializer(VD) &&
+            !S.isDecoratedWithSyclAttribute<SYCLGlobalVariableAllowedAttr>(
```

Comment:
Is there a reason we didn't just do - 

```
const RecordDecl *RD = VD->getType()->getAsRecordDecl();
if (RD && RD->hasAttr<SYCLGlobalVariableAllowedAttr>())
```
 I'm only asking because we've refactored a lot of SYCL attribute code in recent years where we removed a lot of attribute handling code for SYCL only attributes (and used whatever community clang does for their attributes instead) @AaronBallman @smanna12 since you were involved in this, I would like your thoughts on this.

Reply:
There is. At some places in sema the record returned by `variable->gettype()->getasrecrddecl()` doesn’t yet have the attribute on it (I assume that is because templates are not instantiated yet, since at the codegen level it does have it). So, we have to check the template declaration for attribute. I’ve met it and added the check that is re-used by `isDecoratedWithSyclAttribute` function while I was working on integration footer support for device global.